src/DockComponentsFactory.cpp

```diff
@@ -0,0 +1,68 @@
+//============================================================================
+/// \file   DockComponentsFactory.cpp
+/// \author Uwe Kindler
+/// \date   10.02.2020
+/// \brief  Implementation of DockComponentsFactory
+//============================================================================
+
+//============================================================================
+//                                   INCLUDES
+//============================================================================
+#include <DockComponentsFactory.h>
+#include <QScopedPointer>
+
+#include "DockWidgetTab.h"
+#include "DockAreaTabBar.h"
+#include "DockAreaTitleBar.h"
+#include "DockWidget.h"
+#include "DockAreaWidget.h"
+
+namespace ads
+{
+static QScopedPointer<CDockComponentsFactory> DefaultFactory(new CDockComponentsFactory());
+
+
+//============================================================================
+CDockWidgetTab* CDockComponentsFactory::createDockWidgetTab(CDockWidget* DockWidget) const
+{
+	return new CDockWidgetTab(DockWidget);
+}
+
+
+//============================================================================
+CDockAreaTabBar* CDockComponentsFactory::createDockAreaTabBar(CDockAreaWidget* DockArea) const
+{
+	return new CDockAreaTabBar(DockArea);
+}
+
+
+//============================================================================
+CDockAreaTitleBar* CDockComponentsFactory::createDockAreaTitleBar(CDockAreaWidget* DockArea) const
+{
+	return new CDockAreaTitleBar(DockArea);
+}
+
+
+//============================================================================
+const CDockComponentsFactory* CDockComponentsFactory::factory()
+{
+	return DefaultFactory.data();
+}
+
+
+//============================================================================
+void CDockComponentsFactory::setFactory(CDockComponentsFactory* Factory)
+{
+	DefaultFactory.reset(Factory);
+}
+
+
+//============================================================================
+void CDockComponentsFactory::resetDefaultFactory()
+{
+	DefaultFactory.reset(new CDockComponentsFactory());
+}
```

Comment:
For me this works, but I would expect that each one of these components had a minimal interface, so we can create a completely different thing from scratch.
For example, I wouldn't mind using the default Qt tab bar. I could inherit `CDockAreaTabBar` and implement my own layout with my own widgets.
```cpp
class IDockAreaTabBar : public QFrame {
// abstract class
public:
     virtual CDockWidgetTab* dockWidgetTab() const = 0;
public slots:
     virtual void markMenuOutdated() = 0; // example
}

class CDockAreaTabBar : IDockAreaTabBar {
// default implementation
}
```